File: Controllers/MenuController.cs
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Sofranet.Data;
using Sofranet.Models;
using Sofranet.Models.ViewModels;
using Sofranet.Services;

namespace Sofranet.Controllers;

public class MenuController : Controller
{
    private readonly ApplicationDbContext _db;
    private readonly UserManager<ApplicationUser> _userManager;
    private readonly ILocationService _location;
    private const int PageSize = 12;

    public MenuController(
        ApplicationDbContext db,
        UserManager<ApplicationUser> userManager,
        ILocationService location)
    {
        _db = db;
        _userManager = userManager;
        _location = location;
    }

    public async Task<IActionResult> Index(string? caterer, decimal? minPrice, decimal? maxPrice,
        int radius = 10, int page = 1)
    {
        if (page < 1) page = 1;
        if (radius < 1) radius = 10;

        // Lokasyon-bazlı caterer filtresi (sadece User rolü için)
        List<CatererDistance>? nearbyCaterers = null;
        double? userLat = null, userLng = null;
        bool isUserRole = false;
        bool locationAvailable = false;

        if (User.Identity?.IsAuthenticated == true)
        {
            var currentUser = await _userManager.GetUserAsync(User);
            if (currentUser != null && await _userManager.IsInRoleAsync(currentUser, "User"))
            {
                isUserRole = true;
                if (currentUser.Latitude.HasValue && currentUser.Longitude.HasValue)
                {
                    locationAvailable = true;
                    userLat = currentUser.Latitude;
                    userLng = currentUser.Longitude;

                    var catererRole = await _db.Roles.FirstOrDefaultAsync(r => r.Name == "Caterer");
                    var allCaterers = new List<CatererLocation>();
                    if (catererRole != null)
                    {
                        allCaterers = await (from u in _db.Users
                                             join ur in _db.UserRoles on u.Id equals ur.UserId
                                             where ur.RoleId == catererRole.Id
                                                   && u.Latitude.HasValue && u.Longitude.HasValue
                                             select new CatererLocation
                                             {
                                                 Id = u.Id,
                                                 Name = u.FullName,
                                                 Address = u.Address,
                                                 Latitude = u.Latitude!.Value,
                                                 Longitude = u.Longitude!.Value
                                             }).ToListAsync();
                    }
                    nearbyCaterers = await _location.FilterCaterersByDistanceAsync(
                        allCaterers, userLat.Value, userLng.Value, radius);
                }
            }
        }

        var query = _db.MenuItems
            .Include(m => m.Caterer)
            .Where(m => m.IsAvailable);

        if (!string.IsNullOrWhiteSpace(caterer))
            query = query.Where(m => m.CatererId == caterer);

        if (minPrice.HasValue)
            query = query.Where(m => m.Price >= minPrice.Value);

        if (maxPrice.HasValue)
            query = query.Where(m => m.Price <= maxPrice.Value);

        if (nearbyCaterers != null)
        {
            var nearbyIds = nearbyCaterers.Select(c => c.Id).ToList();
            if (nearbyIds.Count == 0)
            {
                query = query.Where(m => false);
            }
            else
            {
                query = query.Where(m => nearbyIds.Contains(m.CatererId));
            }
        }

        var total = await query.CountAsync();
        var items = await query
            .OrderByDescending(m => m.CreatedAt)
            .Skip((page - 1) * PageSize)
            .Take(PageSize)
            .ToListAsync();

        // Caterer dropdown
        var catererRoleRow = await _db.Roles.FirstOrDefaultAsync(r => r.Name == "Caterer");
        var catererList = new List<dynamic>();
        if (catererRoleRow != null)
        {
            catererList = await (from u in _db.Users
                                 join ur in _db.UserRoles on u.Id equals ur.UserId
                                 where ur.RoleId == catererRoleRow.Id
                                 orderby u.FullName
                                 select new { u.Id, u.FullName }).ToListAsync<dynamic>();
        }

        // Average rating'ler
        var itemList = new List<MenuListItemViewModel>();
        foreach (var item in items)
        {
            var menuRatings = await _db.Ratings
                .Where(r => r.MenuItemId == item.Id && r.Order!.Status == "completed")
                .Select(r => (decimal)r.MenuRating)
                .ToListAsync();
            decimal? menuAvg = menuRatings.Count > 0 ? menuRatings.Average() : null;

            var catererRatings = await _db.Ratings
                .Where(r => r.CatererId == item.CatererId && r.Order!.Status == "completed")
                .Select(r => (decimal)r.CatererRating)
                .ToListAsync();
            decimal? catererAvg = catererRatings.Count > 0 ? catererRatings.Average() : null;

            itemList.Add(new MenuListItemViewModel
            {
                Item = item,
                CatererName = item.Caterer?.FullName ?? "-",
                MenuAvg = menuAvg,
                RatingCount = menuRatings.Count,
                CatererAvg = catererAvg
            });
        }

        ViewBag.Caterers = catererList;
        ViewBag.SelectedCaterer = caterer;
        ViewBag.MinPrice = minPrice;
        ViewBag.MaxPrice = maxPrice;
        ViewBag.CurrentPage = page;
        ViewBag.TotalPages = (int)Math.Ceiling(total / (double)PageSize);
        ViewBag.Total = total;

        ViewBag.IsUserRole = isUserRole;
        ViewBag.LocationAvailable = locationAvailable;
        ViewBag.UserLat = userLat;
        ViewBag.UserLng = userLng;
        ViewBag.Radius = radius;
        ViewBag.CatererDistances = nearbyCaterers?.ToDictionary(c => c.Id, c => c.DistanceKm)
            ?? new Dictionary<string, double>();
        ViewBag.MapCaterers = nearbyCaterers ?? new List<CatererDistance>();
        ViewBag.NeedsMaps = locationAvailable;
        ViewBag.GoogleMapsKey = _location.GetApiKey();

        ViewData["Title"] = "Menüler";
        return View(itemList);
    }

    public async Task<IActionResult> Detail(int id)
    {
        var item = await _db.MenuItems
            .Include(m => m.Caterer)
            .Include(m => m.OptionGroups).ThenInclude(g => g.Options)
            .Include(m => m.RemovableIngredients)
            .AsSplitQuery()
            .FirstOrDefaultAsync(m => m.Id == id && m.IsAvailable);

        if (item == null) return NotFound();

        var menuRatings = await _db.Ratings
            .Where(r => r.MenuItemId == id && r.Order!.Status == "completed")
            .Select(r => (decimal)r.MenuRating)
            .ToListAsync();
        decimal? menuAvg = menuRatings.Count > 0 ? menuRatings.Average() : null;

        var catererRatings = await _db.Ratings
            .Where(r => r.CatererId == item.CatererId && r.Order!.Status == "completed")
            .Select(r => (decimal)r.CatererRating)
            .ToListAsync();
        decimal? catererAvg = catererRatings.Count > 0 ? catererRatings.Average() : null;

        var recentComments = await _db.Ratings
            .Include(r => r.User)
            .Where(r => r.MenuItemId == id
                && r.Order!.Status == "completed"
                && r.Comment != null
                && r.Comment != "")
            .OrderByDescending(r => r.CreatedAt)
            .Take(5)
            .Select(r => new RecentCommentViewModel
            {
                UserName = r.User!.FullName,
                MenuRating = r.MenuRating,
                Comment = r.Comment,
                CreatedAt = r.CreatedAt
            })
            .ToListAsync();

        var vm = new MenuDetailViewModel
        {
            Item = item,
            CatererName = item.Caterer?.FullName ?? "-",
            CatererAddress = item.Caterer?.Address,
            CatererLat = item.Caterer?.Latitude,
            CatererLng = item.Caterer?.Longitude,
            MenuAvg = menuAvg,
            RatingCount = menuRatings.Count,
            CatererAvg = catererAvg,
            Groups = item.OptionGroups.OrderBy(g => g.Id).ToList(),
            Removables = item.RemovableIngredients.OrderBy(r => r.Id).ToList(),
            RecentComments = recentComments
        };

        ViewBag.GoogleMapsKey = _location.GetApiKey();
        ViewData["Title"] = item.Name;
        return View(vm);
    }
}

File: controllers/menuController.js
// menu controller - herkese acik menu listesi
const db = require('../config/db');
const locationService = require('../services/locationService');
const { getPagination } = require('../utils/pagination');

// NOT: Distance Matrix API her cagrida quota tuketiyor. Haversine pre-filter
// (services/locationService.js icindeki filterCaterersByDistance) sayesinde
// sadece olasi adaylar Matrix'e gidiyor - bu sekilde quota israfini onluyoruz.
async function list(req, res) {
  try {
    const filters = {
      caterer: req.query.caterer || '',
      minPrice: req.query.minPrice || '',
      maxPrice: req.query.maxPrice || '',
      radius: req.query.radius || '10'
    };

    // konum filtresi sadece logged-in 'user' icin
    let userLoc = null;
    let locationAvailable = false;
    let nearbyCatererIds = null;
    let catererDistances = {}; // id -> distance (km)
    let mapCaterers = []; // harita icin lightweight obje

    const sessUser = req.session.user;
    if (sessUser && sessUser.role === 'user') {
      const u = await db.query(
        'SELECT latitude, longitude FROM users WHERE id = $1',
        [sessUser.id]
      );
      if (u.rows.length && u.rows[0].latitude != null && u.rows[0].longitude != null) {
        userLoc = {
          lat: Number(u.rows[0].latitude),
          lng: Number(u.rows[0].longitude)
        };
        locationAvailable = true;

        const radiusKm = Math.max(1, parseInt(filters.radius, 10) || 10);

        // tum caterer'lari cek (konumu olanlar)
        const catRes = await db.query(
          "SELECT id, name, latitude, longitude, address FROM users WHERE role = 'caterer' AND latitude IS NOT NULL AND longitude IS NOT NULL"
        );

        const nearby = await locationService.filterCaterersByDistance(
          catRes.rows, userLoc, radiusKm
        );

        nearbyCatererIds = nearby.map((c) => c.id);
        for (const c of nearby) {
          catererDistances[c.id] = c.distance;
        }
        mapCaterers = nearby.map((c) => ({
          id: c.id,
          name: c.name,
          lat: Number(c.latitude),
          lng: Number(c.longitude),
          address: c.address,
          distance: c.distance
        }));
      }
    }

    // dinamik WHERE
    const where = ['mi.is_available = TRUE'];
    const params = [];

    if (filters.caterer && !isNaN(parseInt(filters.caterer, 10))) {
      params.push(parseInt(filters.caterer, 10));
      where.push('mi.caterer_id = $' + params.length);
    }
    if (filters.minPrice !== '' && !isNaN(parseFloat(filters.minPrice))) {
      params.push(parseFloat(filters.minPrice));
      where.push('mi.price >= $' + params.length);
    }
    if (filters.maxPrice !== '' && !isNaN(parseFloat(filters.maxPrice))) {
      params.push(parseFloat(filters.maxPrice));
      where.push('mi.price <= $' + params.length);
    }

    // lokasyon filtresi (sadece nearbyCatererIds varsa)
    if (nearbyCatererIds !== null) {
      if (nearbyCatererIds.length === 0) {
        // hicbir yakın caterer yok -> hicbir item dondur
        where.push('1 = 0');
      } else {
        const placeholders = nearbyCatererIds.map((_, i) => '$' + (params.length + i + 1));
        params.push(...nearbyCatererIds);
        where.push('mi.caterer_id IN (' + placeholders.join(',') + ')');
      }
    }

    const whereSql = where.join(' AND ');
    const page = parseInt(req.query.page, 10) || 1;

    // toplam say
    const countRes = await db.query(
      `SELECT COUNT(*) FROM menu_items mi
       JOIN users u ON u.id = mi.caterer_id
       WHERE ${whereSql}`,
      params
    );
    const pag = getPagination(page, 12, parseInt(countRes.rows[0].count, 10));

    params.push(pag.limit);
    params.push(pag.offset);
    const sql = `
      SELECT mi.*, u.name as caterer_name, u.id as caterer_id,
        (SELECT AVG(r.menu_rating)::numeric(3,2) FROM ratings r
           JOIN orders o ON o.id = r.order_id
           WHERE r.menu_item_id = mi.id AND o.status = 'completed') as menu_avg,
        (SELECT COUNT(*) FROM ratings r
           JOIN orders o ON o.id = r.order_id
           WHERE r.menu_item_id = mi.id AND o.status = 'completed') as menu_rating_count,
        (SELECT AVG(r.caterer_rating)::numeric(3,2) FROM ratings r
           JOIN orders o ON o.id = r.order_id
           WHERE r.caterer_id = mi.caterer_id AND o.status = 'completed') as caterer_avg
      FROM menu_items mi
      JOIN users u ON u.id = mi.caterer_id
      WHERE ${whereSql}
      ORDER BY mi.created_at DESC
      LIMIT $${params.length - 1} OFFSET $${params.length}
    `;

    const result = await db.query(sql, params);

    const catListRes = await db.query(
      "SELECT id, name FROM users WHERE role = 'caterer' ORDER BY name"
    );

    res.render('user/menu', {
      title: 'Menüler',
      items: result.rows,
      caterers: catListRes.rows,
      filters,
      locationAvailable,
      catererDistances,
      isUserRole: sessUser && sessUser.role === 'user',
      needsMaps: locationAvailable,
      userLat: userLoc ? userLoc.lat : null,
      userLng: userLoc ? userLoc.lng : null,
      mapCaterers,
      pagination: pag,
      currentQuery: req.query
    });
  } catch (err) {
    console.error('menu list error:', err);
    res.status(500).send('Menü listesi yüklenemedi');
  }
}

async function detail(req, res) {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) {
      return res.status(404).send('Menü bulunamadi');
    }

    const itemRes = await db.query(
      `SELECT mi.*, u.name as caterer_name, u.id as caterer_owner_id, u.address as caterer_address,
         u.latitude as caterer_latitude, u.longitude as caterer_longitude,
         (SELECT AVG(r.menu_rating)::numeric(3,2) FROM ratings r
            JOIN orders o ON o.id = r.order_id
            WHERE r.menu_item_id = mi.id AND o.status = 'completed') as menu_avg,
         (SELECT COUNT(*) FROM ratings r
            JOIN orders o ON o.id = r.order_id
            WHERE r.menu_item_id = mi.id AND o.status = 'completed') as menu_rating_count,
         (SELECT AVG(r.caterer_rating)::numeric(3,2) FROM ratings r
            JOIN orders o ON o.id = r.order_id
            WHERE r.caterer_id = mi.caterer_id AND o.status = 'completed') as caterer_avg
       FROM menu_items mi
       JOIN users u ON u.id = mi.caterer_id
       WHERE mi.id = $1 AND mi.is_available = TRUE`,
      [id]
    );

    if (itemRes.rows.length === 0) {
      return res.status(404).send('Menü bulunamadi');
    }
    const item = itemRes.rows[0];

    const groupsRes = await db.query(
      'SELECT * FROM option_groups WHERE menu_item_id = $1 ORDER BY id',
      [id]
    );
    const groups = groupsRes.rows;
    for (const g of groups) {
      const optsRes = await db.query(
        'SELECT * FROM options WHERE group_id = $1 ORDER BY id',
        [g.id]
      );
      g.options = optsRes.rows;
    }

    const removablesRes = await db.query(
      'SELECT * FROM removable_ingredients WHERE menu_item_id = $1 ORDER BY id',
      [id]
    );

    const commentsRes = await db.query(
      `SELECT r.menu_rating, r.comment, r.created_at, u.name as user_name
       FROM ratings r
       JOIN users u ON u.id = r.user_id
       JOIN orders o ON o.id = r.order_id
       WHERE r.menu_item_id = $1 AND o.status = 'completed'
         AND r.comment IS NOT NULL AND r.comment != ''
       ORDER BY r.created_at DESC LIMIT 5`,
      [id]
    );

    res.render('user/menu-detail', {
      title: item.name,
      item,
      groups,
      removables: removablesRes.rows,
      recentComments: commentsRes.rows,
      needsMaps: true
    });
  } catch (err) {
    console.error('menu detail error:', err);
    res.status(500).send('Detay yüklenemedi');
  }
}

module.exports = { list, detail };
